src/interfaces/chain.cpp

```diff
@@ -158,7 +158,7 @@ class NotificationsProxy : public CValidationInterface
     {
         m_notifications->transactionAddedToMempool(tx);
     }
-    void TransactionRemovedFromMempool(const CTransactionRef& tx) override
+    void TransactionRemovedFromMempool(const CTransactionRef& tx, bool fIsConflicted) override
```

Comment:
Maybe even just pass the original removal reason to the sink? Compressing `REPLACED` into a boolean called `conflicted` is not wrong, but maybe not the most straightforward either.

Reply:
Yes I hesitated at first to call it is_conflicted but it's still an implementation of a `CValidationInterface`  method so `fIsConflicted` is more reasonable so `is_ibd` should be `fInitialDownload` ?

@MarcoFalke, you suggest doing the filtering in `NotificationsProxy` ? Shouldn't this be as lean as possible and just call notifications handler ? My motivation was avoiding linking mempool code in wallet but didn't think specifically about this point.

Reply:
I don't think you need to link any mempool code when you want access to the enum RemovalReason in the wallet.

Right? cc @ryanofsky

Reply:
I restrained conflict to block only, not including anymore mempool replacement which means API is far-more straightforward as wallet conflict == mempool conflict.

If we really  want to use consistent reason codes among clients, we may pull them out of txmempool.h in some node/*.h ?

Reply:
re: https://github.com/bitcoin/bitcoin/pull/18600#discussion_r412589720

> I restrained conflict to block only

Thanks, this seems ok now that the definition of conflicted here matches up with the MemPoolRemovalReason definition. I do think the API would be more useful and clear if it just passed the existing MemPoolRemovalReason value instead introducing a new way to communicate the same information. But thanks for resolving the original concern!

> If we really want to use consistent reason codes among clients, we may pull them out of txmempool.h in some node/*.h ?

I suggested moving it to `src/interfaces/` in https://github.com/bitcoin/bitcoin/pull/18600#discussion_r408516146 because `src/interfaces/` is meant to be a home for types used at the node/wallet boundary. But `src/node/` would be also ok, and leaving it in `txmempool.h` is perfectly fine. I would leave it in `txmempool.h` if using it in this PR, because this PR is already doing a lot of things. The current source code organization is not ideal and is gradually being improved, so design of the `interfaces::Chain::Notifications::transactionRemovedFromMempool` API shouldn't be decided based on what source file an enum definition is located in. The design should try to be useful and consistent and clear, and source code organization should follow the design instead of the other way around

Reply:
Yes I agree on source code organization following the design. I think `MemPoolRemovalReason` may be dry-ed up in itself, IIRC it's not used beyond `::CONFLICTED`. But outside PR scope.